src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -0,0 +1,285 @@
+                        return;
+                }
+            }
+        }
+
+        private protected void ReadVarIntStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<BigInteger> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            int numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+
+            while (!bufferReader.Complete)
+            {
+                var bigInt = ReadBigVarInt(ref bufferReader);
+
+                if (!bigInt.HasValue)
+                    return;
+
+                if (numSkipped++ < stream.Positions.ValuesToSkip)
+                    continue;
+
+                outputValues[numValuesRead++] = bigInt.Value;
+
+                if (numValuesRead >= outputValues.Length)
+                    return;
+            }
+        }
+
+        private protected async Task<int> GetByteRangeAsync(StreamDetail stream, Memory<byte> output)
+        {
+            if (stream != null)
+            {
+                // If last range matches, the previous data will already be buffered.
+                // We can return only the length without requesting the same bytes.
+
+                if (stream.Range != _lastRange)
+                    _lastRangeLength = await _byteRangeProvider.GetRangeAsync(output.Slice(0, stream.Range.Length), stream.Range.Offset);
+
+                _lastRange = stream.Range;
+            }
+
+            return _lastRangeLength;
+        }
+
+        private protected void DecompressByteRange(StreamDetail stream, ReadOnlySpan<byte> compressedInput, Span<byte> decompressedOutput, ref int decompressedLength)
+        {
+            decompressedLength = 0;
+
+            if (stream != null)
+                decompressedLength = StreamData.Decompress(compressedInput.Slice(0, stream.Range.Length), decompressedOutput, _context.CompressionKind);
+        }
+
+        /// <summary>
+        /// Applies the offset position into the decompressed data.
+        /// </summary>
+        private protected ReadOnlySpan<byte> ResizeBuffer(StreamDetail stream, ReadOnlySpan<byte> decompressedBuffer, int decompressedBufferLength)
+        {
+            var rowEntryLength = decompressedBufferLength - stream.Positions.RowEntryOffset;
+
+            return decompressedBuffer.Slice(stream.Positions.RowEntryOffset, rowEntryLength);
+        }
+
+        private protected double BigIntegerToDouble(BigInteger numerator, long scale)
+            => (double)BigIntegerToDecimal(numerator, scale);
+
+        private protected decimal BigIntegerToDecimal(BigInteger numerator, long scale)
+        {
+            if (scale < 0 || scale > 255)
+                throw new OverflowException("Scale must be positive number");
+
+            var decNumerator = (decimal)numerator; //This will throw for an overflow or underflow
```

Comment:
Yes. I didn't check but the `BigInteger` may have some implicit converter checks for that. It will throw an `OverflowException: Value was either too large or too small for a Decimal.`.

Sample can be seen [here](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwLACgACAGABJgRgDoA5CAWwFMwBLAYwGcBuDDAIRoHMBJKYSztVxQK1AIbAA9mFwBeYZQDuuDjz4DqACjiU6NcmIA2RALJiAHgDUjESgEpciXARboAkBgBuYmTrpkqMAlpOVxtXX0jOxFA4LAmIA==).